_books/ion-1-1/src/macros/system_macros.md

```diff
@@ -0,0 +1,398 @@
+(:make_list (1 2))            => [1, 2]
+(:make_list (1 2) [3, 4])     => [1, 2, 3, 4]
+(:make_list ((1 2)) [[3, 4]]) => [(1 2), [3, 4]]
+```
+
+#### `make_sexp`
+
+```ion
+(macro make_sexp (sequences*) ( (.flatten sequences) ))
+```
+
+Like `make_list` but produces a sexp.
+This is the only way to produce an S-expression from a template: unlike lists, S-expressions in
+templates are not quasi-literals.
+
+```ion
+(:make_sexp)                  => ()
+(:make_sexp (1 2))            => (1 2)
+(:make_sexp (1 2) [3, 4])     => (1 2 3 4)
+(:make_sexp ((1 2)) [[3, 4]]) => ((1 2) [3, 4])
+```
+
+#### `make_struct`
+
+```ion
+(macro make_struct (structs*) /* Not representable in TDL */)
+```
+
+Produces a non-null, unannotated struct by combining the fields of any number of non-null structs.
+
+```ion
+(:make_struct)    => {}
+(:make_struct
+  {k1: 1, k2: 2}
+  {k3:3}
+  {k4: 4})        =>  {k1:1, k2:2, k3:3, k4:4}
+```
+
+#### `make_decimal`
+
+```ion
+(macro make_decimal (flex_int::coefficient flex_int::exponent) /* Not representable in TDL */)
+```
+
+This is no more compact than the regular binary encoding for decimals.
+However, it can be used in conjunction with other macros, for example, to represent fixed-point numbers.
+
+```ion
+(macro usd (cents) (annotate (literal USD) (make_decimal cents -2))
+
+(:usd 199) =>  USD::1.99
+```
+
+#### `make_timestamp`
+
+```ion
+(macro make_timestamp (uint16::year
+                       uint8::month?
+                       uint8::day?
+                       uint8::hour?
+                       uint8::minute?
+                       /*decimal*/ second?
+                       int16::offset_minutes?) /* Not representable in TDL */)
+```
+Produces a non-null, unannotated timestamp at various levels of precision.
+When `offset` is absent, the result has unknown local offset; offset `0` denotes UTC.
+The arguments to this macro may not be any null value.
+
+> [!NOTE]
+> TODO [ion-docs#256](https://github.com/amazon-ion/ion-docs/issues/256) Reconsider offset semantics, perhaps default should be UTC.
+
+Example:
+
+```ion
+(macro ts_today 
+       (uint8::hour uint8::minute uint32::seconds_millis)
+       (.make_timestamp
+         2022
+         4
+         28
+         hour
+         minute
+         (.make_decimal seconds_millis -3) 0))
+```
+
+### Encoding Utility Macros
+
+#### `repeat`
+
+The `repeat` system macro can be used for efficient run-length encoding.
+
+```ion
+(macro repeat (n! value+) /* Not representable in TDL */)
+```
+Produces a stream that repeats the specified `value` expression(s) `n` times.
+
+```ion
+(:repeat 5 0) => 0 0 0 0 0
+(:repeat 2 true false) => true false true false
+```
+
+#### `delta`
+
+> [!NOTE]
+> 🚧 Name still TBD 🚧
+
+The `delta` system macro can be used for directed delta encoding.
+
+```ion
+(macro delta (flex_int::initial! flex_int::deltas+) /* Not representable in TDL */)
+```
+
+Example:
+```ion
+(:delta 10 1 2 3 -4) => 11 13 16 12
+```
+
+#### `sum`
```

Comment:
I'm struggling to remember whether we had a use case for `sum` besides biased integers (which `delta` enables and more)

Reply:
I don't remember the use case either, but `sum` only produces one integer vs `delta` producing `N-1` integers, so `delta` can't entirely replace `sum`.

Reply:
I believe we started with `sum` to enable biased integers but then realized `delta` would achieve that _and_ delta encoding.

For serialization, I can't think of a distinct use case for `sum`. Writing a sequence of numbers to mean their total will always take more bytes.